modules/ismp/clients/polygon/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+	/// Milestone update
+	pub milestone_update: Option<MilestoneUpdate>,
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct MilestoneUpdate {
+	/// EVM block header for the milestone's end block
+	pub evm_header: geth_primitives::CodecHeader,
+	/// Milestone number
+	pub milestone_number: u64,
+	/// ICS23 proof for the milestone inclusion
+	pub ics23_state_proof: ICS23Proof,
+	// Milestone
+	pub milestone: Milestone,
+	/// Untrusted header app hash for verification
+	pub untrusted_header_app_hash: Vec<u8>,
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct ICS23Proof {
+	/// Proof bytes
+	pub proof: Vec<u8>,
+	/// Key
+	pub key: Vec<u8>,
+	/// Value
+	pub value: Vec<u8>,
+}
+
+/// The trusted consensus state for Polygon
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct ConsensusState {
+	///  Codec Trusted Tendermint state
+	pub tendermint_state: Vec<u8>,
+	/// Last finalized Polygon block number
+	pub last_finalized_block: u64,
+	/// Last finalized Polygon block hash
+	pub last_finalized_hash: Vec<u8>,
+}
+
+pub const POLYGON_CONSENSUS_CLIENT_ID: ConsensusClientId = *b"PLGN"; // TODO: Change to Polygon Consensus ID
+
+const POLYGON_CHAIN_ID: u32 = 137;
+const POLYGON_TESTNET_CHAIN_ID: u32 = 80002;
+
+pub struct PolygonClient<H: IsmpHost, T: HostExecutiveConfig>(core::marker::PhantomData<(H, T)>);
+
+impl<H: IsmpHost, T: HostExecutiveConfig> Default for PolygonClient<H, T> {
+	fn default() -> Self {
+		Self(core::marker::PhantomData)
+	}
+}
+
+impl<H: IsmpHost + Send + Sync + Default + 'static, T: HostExecutiveConfig> ConsensusClient
+	for PolygonClient<H, T>
+{
+	fn verify_consensus(
+		&self,
+		_host: &dyn IsmpHost,
+		_consensus_state_id: ConsensusStateId,
+		trusted_consensus_state: Vec<u8>,
+		proof: Vec<u8>,
+	) -> Result<(Vec<u8>, VerifiedCommitments), ismp::error::Error> {
+		let polygon_consensus_update: PolygonConsensusUpdate = Decode::decode(&mut &proof[..])
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?;
+
+		let consensus_state = ConsensusState::decode(&mut &trusted_consensus_state[..])
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?;
+
+		let milestone_update = polygon_consensus_update.milestone_update.clone();
+
+		if let Some(milestone_update_ref) = &polygon_consensus_update.milestone_update {
+			// Check milestone_number == milestone.end_block and both match evm_header.number
+			let end_block =
+				milestone_update_ref.milestone.end_block.parse::<u64>().map_err(|_| {
+					ismp::error::Error::Custom("Invalid milestone.end_block".to_string())
+				})?;
+			let evm_block = milestone_update_ref.evm_header.number.low_u64();
+			if milestone_update_ref.milestone_number != end_block || end_block != evm_block {
+				return Err(ismp::error::Error::Custom(
+					"Milestone number, end_block, and EVM header number do not match".to_string(),
+				));
+			}
+
+			if milestone_update_ref.milestone_number <= consensus_state.last_finalized_block {
```

Comment:
Milestone number is not related to block numbers, it should not be compared to anything.

The only comparisons you need in this verifier are

milestone.hash == evm_header_hash
milestone.end_block == evm_header.number
evm_header.number > consensus_state.last_finalized_block